test/uniport-tests.js

```diff
@@ -0,0 +1,502 @@
+  // port, to determine readiness for tests on the other ports. This Remote's
+  // trace is configured in the normal way, however we can set trace for each of
+  // the test connections on the various ports.
+  remote_trace : false,
+
+  // There's a bug where we can't subscribe to the admin ports requiring
+  // credentials, even though subscribe has Role::USER. `Remote` always issues a
+  // subscribe command after connect(). We can just add the credentials to let
+  // the tests go further to test that admin requests without credentials are
+  // not honoured.
+  add_credentials_to_ws_subscribe: false,
+
+  // WebSocket connections honour admin=no regardless of ip.
+  // Yet http with accepts admin commands from any whitelisted ip.
+  // 127.0.0.1 is hardcoded in cpp code and added to [rpc_admin_allow].
+  expect_http_to_honour_admin_no: process.env.HTTP_HONOUR_ADMIN_NO || true,
+
+  // We can be a bit more exhaustive, but it will cost us 2 seconds, unless
+  // we use `skip_tests_matching` to just `test.skip` them.
+  define_redundant_tests : process.env.REDUNDANT_UNIPORT_TESTS || true,
+
+  skip_tests_matching : [
+    // /redundant/
+
+    // /can not issue/,
+    // /http/
+  ],
+
+  skip_tests_not_matching : [
+    // /wrong password/,
+    // /ws/
+  ],
+
+  skip_ports_not_matching : [
+    // /admin/,
+    // /password/,
+    // /http/
+  ],
+
+  skip_ports_matching : [
+    // /admin/,
+    // /password/,
+    // /http/
+  ]
+}
+
+/* --------------------------------- HELPERS -------------------------------- */
+
+var eachItem = function (o, f) {
+  for (var k in o) {
+    if (o.hasOwnProperty(k)) f(k, o[k], o)
+  }
+}
+
+var prettyj = function (o)
+{
+  return JSON.stringify(o, undefined, 2);
+}
+
+var client_protocols = function(conf) {
+  return conf.protocol.split(',').filter(function(p) {return p != 'peer'});
+}
+
+var opposite_protocol = function (protocol) {
+  switch(protocol)
+  {
+    case 'ws':
+      return 'wss';
+    case 'http':
+      return 'https';
+    case 'wss':
+      return 'ws';
+    case 'https':
+      return 'http';
+    default:
+      throw new Error('unknown protocol '+ protocol);
+  }
+}
+
+var once = function (f) {
+  var alreadyDone = false;
+  return function doneOnce() {
+    // console.log(arguments.callee.caller.toString())
+    if (!alreadyDone) {
+      f();
+    };
+    alreadyDone = true;
+  }
+}
+
+var define_test = testutils.definer_matching(
+    {
+      skip_if_not_match:configurations.skip_tests_not_matching,
+      skip_if_match:configurations.skip_tests_matching
+    },
+    global.test );
+
+var define_suite = testutils.definer_matching(
+    {
+      skip_if_not_match:configurations.skip_ports_not_matching,
+      skip_if_match:configurations.skip_ports_matching
+    },
+    global.suite );
+
+mark_redundant_tests = function(normalizer) {
+  var tested = [];
+
+  return function(test_name, func) {
+    var redundant;
```

Comment:
Tiny quibble: define when you assign:

    var normed = normalizer ? normalizer(test_name) : test_name;
    var redundant = ~tested.indexOf(normed);
    if (redundant)
        test_name += (' (redundant)');
    else
        tested.push(normed);